### PR TITLE
Bdd simplify

### PR DESCRIPTION
**Pull request by marcogario**
Simplifier based on BDD conversion. Provides additional utilities like validation and handling of SMT formulas via boolean abstraction.

### CODE REVIEW

File: pysmt/simplifier.py [hunk 1]
```diff
@@ -664,3 +666,73 @@ def walk_div(self, formula, args, **kwargs):
...
+    def _validate(self, old, new):
+        if self.validate_simplifications:
+            Iff = self.env.formula_manager.Iff
+            is_valid = self.env.factory.is_valid
+            sname = self._validation_sname
+            assert is_valid(Iff(old, new), solver_name=sname ), \
+              "Was: %s \n Obtained: %s\n" % (str(old), str(new))
+
+    def abstract_and_simplify(self, formula):
+        abs_formula = self.walk(formula)
```
Comment (mikand) [inline pysmt/simplifier.py]: Would it make sense to invoke the syntactical simplification beforehand? If an atom simplifies to TRUE or FALSE the BDD could be smaller....

Comment (mikand) [inline pysmt/simplifier.py]: Even if the Simplifier recognizes that two atoms are equivalent the BDD could be reduced...

File: pysmt/simplifier.py [hunk 2]
```diff
@@ -664,3 +666,73 @@ def walk_div(self, formula, args, **kwargs):
         return self.manager.Div(sl, sr)
 
 # EOC Simplifier
+
+
+class BddSimplifier(Simplifier):
```
Comment (mikand) [inline pysmt/simplifier.py]: This class will throw an exception at construction time if pyCUDD is not installed, it is OK for now, but for the next simplifier we would need a factory class...

File: pysmt/walkers/dag.py [hunk 3]
```diff
@@ -70,8 +70,11 @@ def _compute_node_result(self, formula, **kwargs):
             except KeyError:
                 f = self.walk_error
 
-            args = [self.memoization[self._get_key(s, **kwargs)] \
-                    for s in self._get_children(formula)]
+            try:
+                args = [self.memoization[self._get_key(s, **kwargs)] \
+                        for s in self._get_children(formula)]
+            except KeyError as ex:
```
Comment (mikand) [inline pysmt/walkers/dag.py]: This should never happen if the walker is correct, right?

File: pysmt/simplifier.py [hunk 4]
```diff
@@ -664,3 +666,73 @@ def walk_div(self, formula, args, **kwargs):
...
+              "Was: %s \n Obtained: %s\n" % (str(old), str(new))
+
+    def abstract_and_simplify(self, formula):
+        abs_formula = self.walk(formula)
+        abs_res = self.back(self.convert(abs_formula))
+        res = abs_res.substitute(self.ba_map)
+        return res
+
+    def walk_abstract(self, formula, args, **kwargs):
+        type_ = self.get_type(formula)
```
Comment (marcogario) [inline pysmt/simplifier.py]: This is not needed. Here it will always be BOOL

**Comment by marcogario** [comment]
I addressed the comments in the PR. Ready to be reviewed again.

**Pull request merged**